### ISSUE TITLE
Evolve to generic 3Di schematisation api

### ISSUE THREAD

**Issue by leendertvanwolfswinkel** [nens/threedi-modelchecker#214]
Ticket to collect & develop ideas to further develop this package to a generic python API for 3Di schematisations, explicitly made available to users as such

Useful features / convenience functions:
- Create new spatialite with database schema but no data
- Function to get list of paths of referenced rasters

**Comment by Jonasvsl** [comment]
@leendertvanwolfswinkel  I think it can be of great help to look at the hydrolib package that Deltares developed. I expect that we need the same type of functionality as they have now implemented
I think for us the most interesting part is the wirting to model db / sqlite kind of implmentation. This notebook might be a useful example: https://github.com/Deltares/HYDROLIB/blob/main/hydrolib/notebooks/Hydrolib-D-Hydamo_usage_introduction_RRFM.ipynb

And does it also mean that the schematisation editor is going to be the basis for writing the SQLITE?

**Comment by Jonasvsl** [comment]
From @caspervdw  I understand we have a ORM implemented here:
https://github.com/nens/threedi-modelchecker

Preferably this is used to create tools that write to SQLITE
Isn't this already for a large part what users are looking for?

**Comment by leendertvanwolfswinkel** [comment]
Yes. That's why I made this ticket in this repo and used the term "evolve". I couldn't really find a better word.

This repo/package should become the logical starting point for any user that wants to use python to talk to the spatialite, which it now clearly isn't (nobody knows of its existence/possibilities, no documentation, no examples, possibly not a very user friendly api/syntax required, etc..).

**Comment by caspervdw** [comment]
Some ideas that I have about making the ORM public:

**Compatibility**
If you expose the ORM publicly, you have to provide some compatibility guarantees. It is of no use if every release of modelchecker breaks your library. You could pin the modelchecker version but then you will not be compatible with newer sqlite files. And using multiple libraries each with different pinned modelchecker versions is a problem that you can't solve in Python. Your environment must be consistent.

With a schema change, the ORM may change in a backwards incompatible way. We could limit ourselves to non-breaking changes (e.g. never delete a field, never make a nullable field not-nullable, never change defaults) but I think we want to have more flexibility then that?

Another way to guarantee more compatibility is making the API an additional layer on top of the ORM. This is a considerable project. But then we could maintain the library in a backwards compatible way.

So our options are:
1. Accept breaking changes for modelchecker
2. Don't change the sqlite schema in breaking ways
3. Add an additional compatibility layer on top of the ORM

Am I missing something here?

**Comment by leendertvanwolfswinkel** [comment]
I agree, stability and backwards compatibility would also be an important value of this library. 

I prefer an additional layer on top of the ORM, also because the syntax is currently too complex. For example, if I want to get a few global settings, I want to do something like this:

```
schematisation_database = SchematisationDatabase(r"C:\path\to\my.sqlite")
simulation_time_step = schematisation_database.global_settings.simulation_time_step
nr_timesteps = schematisation_database.global_settings.nr_timesteps
```
Instead of (current syntax):

```
database = ThreediDatabase(
    connection_settings={"db_path": sqlite_path.as_posix()},
    db_type="spatialite",
)
try:
    session = database.get_session()
    simulation_timestep, nr_timesteps = Query(
        [GlobalSetting.sim_time_step, GlobalSetting.nr_timesteps]
    ).with_session(session)[0]
finally:
    session.close()
```

**Comment by caspervdw** [comment]
I agree that working with SQLalchemy is quite complex. Together with the compatibility argument I think it is worth spending time on a schematisation API.

As an API it seems logical to follow “threedigrid”. Technically there is already some basic stuff in “threedigrid-builder” (making use of numpy & pygeos)

An additional layer of abstraction would also give an opportunity to combine certain tables into more complex objects (pipe + crosssection definition, channel + crosssection location).

Anyway, this is no evolution of an existing API then. It is really something new.
